src/Installer/PluginInstaller.php

```diff
@@ -2,89 +2,235 @@
+        };
+
+        $messages = [
+            '',
+            '',
+            $wrap(''),
+            $wrap('Action required!'),
+            $wrap(''),
+            $wrap('The CakePHP plugin installer has been changed, please update your'),
+            $wrap('application composer.json file to add the post-autoload-dump hook.'),
+            $wrap('See the changes in https://github.com/cakephp/app/pull/216 for more'),
+            $wrap('info.'),
+            $wrap(''),
+            '',
+            '',
+        ];
+
+        $this->io->write($messages);
+    }
+
+    /**
+     * Called whenever composer (re)generates the autoloader
      *
-     * @param \Composer\Repository\InstalledRepositoryInterface $repo Repository in which to check.
-     * @param \Composer\Package\PackageInterface $package Package instance.
+     * Recreates CakePHP's plugin path map, based on composer information
+     * and available app-plugins.
+     *
+     * @param Event $event the composer event object
+     * @return void
      */
-    public function install(InstalledRepositoryInterface $repo, PackageInterface $package)
+    public static function postAutoloadDump(Event $event)
     {
-        parent::install($repo, $package);
-        $path = $this->getInstallPath($package);
-        $ns = $this->primaryNamespace($package);
-        $this->updateConfig($ns, $path);
+        $composer = $event->getComposer();
+        $config = $composer->getConfig();
+
+        $vendorsDir = $config->get('vendor-dir');
+        $root = dirname($vendorsDir);
+
+        $packages = $composer->getRepositoryManager()->getLocalRepository()->getPackages();
+        $pluginsDir = $root . DIRECTORY_SEPARATOR . 'plugins';
+
+        $plugins = static::determinePlugins($packages, $pluginsDir, $vendorsDir);
+
+        $configFile = static::_configFile($root);
+        static::writeConfigFile($configFile, $plugins);
     }
 
     /**
-     * Updates specific plugin.
+     * Find all plugins available
      *
-     * After the plugin is installed, app's `plugins.php` config file is updated with
-     * plugin namespace to path mapping.
+     * Add all composer packages of type cakephp-plugin, and all plugins located
+     * in the plugins directory to a plugin-name indexed array of paths
      *
-     * @param \Composer\Repository\InstalledRepositoryInterface $repo Repository in which to check.
-     * @param \Composer\Package\PackageInterface $initial Already installed package version.
-     * @param \Composer\Package\PackageInterface $target Updated version.
+     * @param array $packages an array of \Composer\Package\PackageInterface objects
+     * @param string $pluginsDir the path to the plugins dir
+     * @param string $vendorsDir the path to the vendors dir
+     * @return array plugin-name indexed paths to plugins
+     */
+    public static function determinePlugins($packages, $pluginsDir = 'plugins', $vendorsDir = 'vendors')
```

Comment:
Although only used with passing a value for $vendorsDir, default value of `'vendor'` would be a better fit